simulator/src/main/cook/sim/reporting.clj

```diff
@@ -0,0 +1,327 @@
+  "Time from scheduling to finish."
+  [sim-job scheduled-at instances]
+  (let [finished-at (first-finish instances)]
+    (if finished-at
+      (- finished-at scheduled-at))))
+
+(defn job-result
+  [sim-db cook-db [job-id action-at]]
+  (let [sim-job (:actionLog/action (d/entity sim-db [:job/uuid job-id]))
+        cook-job (d/entity cook-db [:job/uuid job-id])
+        instances (:job/instance cook-job)]
+    {:id job-id
+     :requested-at action-at
+     :name (:job/name sim-job)
+     :username (:job/username sim-job)
+     :details (d/touch cook-job)
+     :wait-time (wait-time action-at (first instances))
+     :turnaround (turnaround-time sim-job action-at instances)
+     :overhead (overhead-time sim-job action-at instances)
+     :instance-count (count instances)}))
+
+(defn warn-about-unscheduled
+  [jobs]
+  (let [unscheduled-jobs (remove :wait-time jobs)
+        unscheduled-count (count unscheduled-jobs)]
+    (if (pos? unscheduled-count)
+      (do
+        (println "Warning!  This simulation had " unscheduled-count
+                 " jobs that were never scheduled by Cook.")
+        (println "This could be because you are analyzing before the jobs had a chance to finish...")
+       (println "or it could mean that there's a problem with Cook,")
+       (println "or with the Simulator's ability to connect to Cook.")
+       (doseq [job unscheduled-jobs]
+         (println "Job" (:id job))
+         (println "details from Cook DB:" (:details job)))))))
+
+(defn unfinished-jobs
+  [jobs]
+  (->> jobs (filter :wait-time) (remove :turnaround)))
+
+(defn warn-about-unfinished
+  [jobs]
+  (let [unfinished (unfinished-jobs jobs)
+        unfinished-count (count unfinished)]
+    (if (pos? unfinished-count)
+      (do
+        (println "Warning!  This simulation had " unfinished-count " jobs that were started but never finished.")
+        (println "This makes it so that some creativity must be used in order to present average turnaround/overhead time for the entire sim, or to do job-by-job performance comparisions against another sim.")
+        (doseq [job unfinished]
+          (println "Job" (:id job))
+          ;;(println "details from Cook DB:" (:details job))
+          )))))
+
+(defn summarize-preemption
+  [jobs]
+  (let [preemption-count (comp count
+                               (partial filter :instance/preempted?)
+                               :job/instance :details)
+        preemption-counts (into (sorted-map) (group-by preemption-count jobs))
+        total-preemptions (apply + (map preemption-count jobs))]
+    (println "There were" total-preemptions "total preemptions.")
+    (doseq [[number js] preemption-counts]
+      (println "  " (count js) "jobs preempted" number "times."))
+    total-preemptions))
+
+(defn average-of-metric
+  [jobs metric]
+  (let [metrics (remove nil? (map metric jobs))]
+    (if (-> metrics count pos?)
+      (/ (apply + metrics) (count metrics)))))
+
+
+(defn show-average-wait
```

Comment:
This is in regard to the show-average-* functions below. Average can be misleading. Please include percentiles 25%,50%,75%,95%,99%